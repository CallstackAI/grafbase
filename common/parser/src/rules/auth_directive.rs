use super::visitor::{Visitor, VisitorContext};

use dynaql::ServerError;
use dynaql_parser::types::ConstDirective;
use dynaql_value::ConstValue;

pub const AUTH_DIRECTIVE: &str = "auth";

pub struct AuthDirective;

#[derive(Debug, serde::Serialize, serde::Deserialize)]
struct Auth {
    providers: Vec<AuthProvider>,
}

#[derive(Debug, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "lowercase")]
#[serde(tag = "type")]
#[serde(deny_unknown_fields)]
enum AuthProvider {
    #[serde(rename_all = "camelCase")]
    Oidc {
        issuer: url::Url,
        groups: Option<Vec<String>>,
    },
}

impl<'a> Visitor<'a> for AuthDirective {
    // FIXME: this snippet is parsed, but not enforced by the server, why?
    fn directives(&self) -> String {
        r#"
        directive @auth(providers: [AuthProviderDefinition!]!) on SCHEMA
        input AuthProviderDefinition {
          type: String!
          issuer: String!
        }
        "#
        .to_string()
    }

    fn enter_schema(
        &mut self,
        ctx: &mut VisitorContext<'a>,
        schema_definition: &'a dynaql::Positioned<dynaql_parser::types::SchemaDefinition>,
    ) {
        if let Some(directive) = schema_definition
            .node
            .directives
            .iter()
            .find(|d| d.node.name.node == AUTH_DIRECTIVE)
        {
            match (&directive.node).try_into() as Result<Auth, ServerError> {
                Ok(auth) => {
                    ctx.registry.get_mut().auth = Some(auth.into());
                }
                Err(err) => {
                    ctx.report_error(vec![directive.pos], err.message);
                }
            }
        }
    }
}

impl TryFrom<&ConstDirective> for Auth {
    type Error = ServerError;

    fn try_from(value: &ConstDirective) -> Result<Self, Self::Error> {
        let pos = Some(value.name.pos);

        let arg = match value.get_argument("providers") {
            Some(arg) => match &arg.node {
                ConstValue::List(value) => value,
                _ => return Err(ServerError::new("auth providers must be a list", pos)),
            },
            None => return Err(ServerError::new("auth providers missing", pos)),
        };

        let providers = arg
            .iter()
            .map(AuthProvider::try_from)
            .collect::<Result<_, _>>()
            .map_err(|err| ServerError::new(err.message, pos))?;

        Ok(Auth { providers })
    }
}

impl TryFrom<&ConstValue> for AuthProvider {
    type Error = ServerError;

    fn try_from(value: &ConstValue) -> Result<Self, Self::Error> {
        // We convert the value to JSON to leverage serde for deserialization
        let value = match value {
            ConstValue::Object(_) => value
                .clone()
                .into_json()
                .map_err(|err| ServerError::new(err.to_string(), None))?,
            _ => return Err(ServerError::new("auth provider must be an object", None)),
        };

        let provider: AuthProvider =
            serde_json::from_value(value).map_err(|err| ServerError::new(format!("auth provider: {err}"), None))?;

        Ok(provider)
    }
}

impl From<Auth> for dynaql::Auth {
    fn from(auth: Auth) -> Self {
        Self {
            oidc_providers: auth
                .providers
                .iter()
                .map(|provider| match provider {
                    AuthProvider::Oidc { issuer, groups } => dynaql::OidcProvider {
                        issuer: issuer.clone(),
                        groups: groups.clone(),
                    },
                })
                .collect(),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::AuthDirective;
    use crate::rules::visitor::{visit, VisitorContext};
    use dynaql_parser::parse_schema;
    use pretty_assertions::assert_eq;

    #[test]
    fn test_oidc_basic() {
        let schema = r#"
            schema @auth(providers: [
              { type: "oidc", issuer: "https://my.idp.com" }
            ]) {
              query: Boolean # HACK: make top-level auth directive work
            }
            "#;

        let schema = parse_schema(schema).unwrap();
        let mut ctx = VisitorContext::new(&schema);
        visit(&mut AuthDirective, &mut ctx, &schema);

        assert!(ctx.errors.is_empty());
        assert_eq!(
            ctx.registry.borrow().auth.as_ref().unwrap().oidc_providers,
            vec![dynaql::OidcProvider {
                issuer: url::Url::parse("https://my.idp.com").unwrap(),
                groups: None,
            }]
        );
    }

    #[test]
    fn test_oidc_groups() {
        let schema = r#"
            schema @auth(providers: [
              { type: "oidc", issuer: "https://my.idp.com", groups: ["admin"] }
            ]) {
              query: Boolean
            }
            "#;

        let schema = parse_schema(schema).unwrap();
        let mut ctx = VisitorContext::new(&schema);
        visit(&mut AuthDirective, &mut ctx, &schema);

        assert!(ctx.errors.is_empty());
        assert_eq!(
            ctx.registry.borrow().auth.as_ref().unwrap().oidc_providers,
            vec![dynaql::OidcProvider {
                issuer: url::Url::parse("https://my.idp.com").unwrap(),
                groups: Some(vec!["admin".to_string()]),
            }]
        );
    }

    #[test]
    fn test_oidc_missing_field() {
        let schema = r#"
            schema @auth(providers: [
              { type: "oidc" }
            ]) {
              query: Boolean
            }
            "#;

        let schema = parse_schema(schema).unwrap();
        let mut ctx = VisitorContext::new(&schema);
        visit(&mut AuthDirective, &mut ctx, &schema);

        assert_eq!(ctx.errors.len(), 1);
        assert_eq!(
            ctx.errors.get(0).unwrap().message,
            "auth provider: missing field `issuer`",
        );
    }
}
