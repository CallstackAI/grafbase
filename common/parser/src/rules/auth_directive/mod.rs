use dynaql::{Positioned, ServerError};
use dynaql_parser::types::ConstDirective;

use crate::rules::model_directive::MODEL_DIRECTIVE;
use crate::{Visitor, VisitorContext};

mod config;
mod operations;
mod providers;
mod rules;

use config::AuthConfig;

const AUTH_DIRECTIVE: &str = "auth";

pub struct AuthDirective;

impl AuthDirective {
    pub fn parse(
        ctx: &mut VisitorContext<'_>,
        directives: &[Positioned<ConstDirective>],
        is_global: bool,
    ) -> Result<Option<dynaql::AuthConfig>, ServerError> {
        if let Some(directive) = directives.iter().find(|d| d.node.name.node == AUTH_DIRECTIVE) {
            AuthConfig::from_value(ctx, &directive.node, is_global).map(|auth| Some(dynaql::AuthConfig::from(auth)))
        } else {
            Ok(None)
        }
    }
}

impl<'a> Visitor<'a> for AuthDirective {
    // This snippet is parsed, but not enforced by the server, which is why we
    // don't bother adding detailed types here.
    fn directives(&self) -> String {
        format!("directive @{AUTH_DIRECTIVE} on SCHEMA | OBJECT")
    }

    fn enter_schema(
        &mut self,
        ctx: &mut VisitorContext<'a>,
        schema_definition: &'a dynaql::Positioned<dynaql_parser::types::SchemaDefinition>,
    ) {
        match Self::parse(ctx, &schema_definition.node.directives, true) {
            Ok(Some(auth)) => {
                ctx.registry.get_mut().auth = auth;
            }
            Err(err) => {
                ctx.report_error(err.locations, err.message);
            }
            _ => {}
        }
    }

    // Visit types to check that the auth directive is used correctly. Actual
    // processing happens in the model directive.
    fn enter_type_definition(
        &mut self,
        ctx: &mut VisitorContext<'a>,
        type_definition: &'a dynaql::Positioned<dynaql_parser::types::TypeDefinition>,
    ) {
        if let (Some(auth_directive), false) = (
            type_definition
                .node
                .directives
                .iter()
                .find(|d| d.node.name.node == AUTH_DIRECTIVE),
            type_definition
                .node
                .directives
                .iter()
                .any(|d| d.node.name.node == MODEL_DIRECTIVE),
        ) {
            ctx.report_error(
                vec![auth_directive.pos],
                format!("The @{AUTH_DIRECTIVE} directive can only be used on @{MODEL_DIRECTIVE} types"),
            );
        }
    }
}

#[cfg(test)]
mod tests {
    use std::collections::HashMap;

    use super::*;
    use crate::rules::model_directive::ModelDirective;
    use crate::rules::visitor::visit;
    use dynaql_parser::parse_schema;
    use pretty_assertions::assert_eq;
    use providers::DEFAULT_GROUPS_CLAIM;

    macro_rules! parse_test {
        ($fn_name:ident, $schema:literal, $expect:expr) => {
            parse_test!($fn_name, $schema, HashMap::new(), $expect);
        };
        ($fn_name:ident, $schema:literal, $variables:expr, $expect:expr) => {
            #[test]
            fn $fn_name() {
                let variables = $variables;
                let schema = parse_schema($schema).unwrap();
                let mut ctx = VisitorContext::new_with_variables(&schema, &variables);
                visit(&mut AuthDirective, &mut ctx, &schema);
                visit(&mut ModelDirective, &mut ctx, &schema);

                assert!(ctx.errors.is_empty(), "errors: {:?}", ctx.errors);
                assert_eq!(ctx.registry.borrow().auth, $expect);
            }
        };
    }

    macro_rules! parse_fail {
        ($fn_name:ident, $schema:literal, $err:literal) => {
            parse_fail!($fn_name, $schema, HashMap::new(), $err);
        };
        ($fn_name:ident, $schema:literal, $variables:expr, $err:literal) => {
            #[test]
            fn $fn_name() {
                let variables = $variables;
                let schema = parse_schema($schema).unwrap();
                let mut ctx = VisitorContext::new_with_variables(&schema, &variables);
                visit(&mut AuthDirective, &mut ctx, &schema);
                visit(&mut ModelDirective, &mut ctx, &schema);

                assert_eq!(ctx.errors.len(), 1, "errors: {:?}", ctx.errors);
                assert_eq!(ctx.errors.get(0).unwrap().message, $err);
            }
        };
    }

    parse_test!(
        no_auth_directive,
        r#"
        schema {
          query: Query
        }
        "#,
        Default::default()
    );

    parse_test!(
        anonymous_rule,
        r#"
        schema @auth(
          rules: [ { allow: anonymous } ]
        ){
          query: Query
        }
        "#,
        Default::default()
    );

    parse_fail!(
        anonymous_rule_with_unsupported_ops,
        r#"
        schema @auth(
          rules: [ { allow: anonymous, operations: [read] } ]
        ){
          query: Query
        }
        "#,
        "auth rule: unknown field `operations`, there are no fields"
    );

    parse_test!(
        private_rule,
        r#"
        schema @auth(
          providers: [ { type: oidc, issuer: "https://my.idp.com" } ]
          rules: [ { allow: private } ]
        ){
          query: Query
        }
        "#,
        dynaql::AuthConfig {
            allowed_private_ops: dynaql::Operations::all(),
            oidc_providers: vec![dynaql::OidcProvider {
                issuer: url::Url::parse("https://my.idp.com").unwrap(),
                groups_claim: DEFAULT_GROUPS_CLAIM.to_string(),
            }],
            ..Default::default()
        }
    );

    parse_test!(
        private_rule_with_ops,
        r#"
        schema @auth(
          providers: [ { type: oidc, issuer: "https://my.idp.com" } ]
          rules: [ { allow: private, operations: [create, delete] } ]
        ){
          query: Query
        }
        "#,
        dynaql::AuthConfig {
            allowed_private_ops: dynaql::Operations::CREATE | dynaql::Operations::DELETE,
            oidc_providers: vec![dynaql::OidcProvider {
                issuer: url::Url::parse("https://my.idp.com").unwrap(),
                groups_claim: DEFAULT_GROUPS_CLAIM.to_string(),
            }],
            ..Default::default()
        }
    );

    parse_test!(
        private_rule_with_empty_ops,
        r#"
        schema @auth(
          providers: [ { type: oidc, issuer: "https://my.idp.com" } ]
          rules: [ { allow: private, operations: [] } ]
        ){
          query: Query
        }
        "#,
        dynaql::AuthConfig {
            allowed_private_ops: dynaql::Operations::empty(),
            oidc_providers: vec![dynaql::OidcProvider {
                issuer: url::Url::parse("https://my.idp.com").unwrap(),
                groups_claim: DEFAULT_GROUPS_CLAIM.to_string(),
            }],
            ..Default::default()
        }
    );

    parse_test!(
        groups_rule,
        r#"
        schema @auth(
          providers: [ { type: oidc, issuer: "https://my.idp.com" } ]
          rules: [ { allow: groups, groups: ["admin", "moderator"] } ],
        ){
          query: Query
        }
        "#,
        dynaql::AuthConfig {
            allowed_group_ops: HashMap::from_iter(vec![
                ("admin".to_string(), dynaql::Operations::all()),
                ("moderator".to_string(), dynaql::Operations::all()),
            ]),
            oidc_providers: vec![dynaql::OidcProvider {
                issuer: url::Url::parse("https://my.idp.com").unwrap(),
                groups_claim: DEFAULT_GROUPS_CLAIM.to_string(),
            }],
            ..Default::default()
        }
    );

    parse_test!(
        groups_rule_with_ops,
        r#"
        schema @auth(
          providers: [ { type: oidc, issuer: "https://my.idp.com" } ]
          rules: [
            { allow: groups, groups: ["admin"] }
            { allow: groups, groups: ["moderator", "editor"], operations: ["get", "list"] }
          ],
        ){
          query: Query
        }
        "#,
        dynaql::AuthConfig {
            allowed_group_ops: HashMap::from_iter(vec![
                ("admin".to_string(), dynaql::Operations::all()),
                (
                    "moderator".to_string(),
                    dynaql::Operations::GET | dynaql::Operations::LIST
                ),
                ("editor".to_string(), dynaql::Operations::GET | dynaql::Operations::LIST)
            ]),
            oidc_providers: vec![dynaql::OidcProvider {
                issuer: url::Url::parse("https://my.idp.com").unwrap(),
                groups_claim: DEFAULT_GROUPS_CLAIM.to_string(),
            }],
            ..Default::default()
        }
    );

    parse_fail!(
        groups_rule_with_duplicate_group,
        r#"
        schema @auth(
          rules: [ { allow: groups, groups: ["A", "B", "B"] } ],
        ){
          query: Query
        }
        "#,
        "auth rule: invalid entry: found duplicate value"
    );

    parse_fail!(
        groups_rule_with_null_group,
        r#"
        schema @auth(
          rules: [ { allow: groups, groups: ["A", null, "B"] } ],
        ){
          query: Query
        }
        "#,
        "auth rule: invalid type: null, expected a string"
    );

    parse_fail!(
        groups_rule_with_null_groups,
        r#"
        schema @auth(
          rules: [ { allow: groups, groups: null } ],
        ){
          query: Query
        }
        "#,
        "auth rule: invalid type: null, expected a sequence"
    );

    parse_fail!(
        groups_rule_with_empty_groups,
        r#"
        schema @auth(
          rules: [ { allow: groups, groups: [] } ],
        ){
          query: Query
        }
        "#,
        "groups must be a non-empty list"
    );

    parse_test!(
        owner_rule,
        r#"
        schema @auth(
          providers: [ { type: oidc, issuer: "https://my.idp.com" } ]
          rules: [ { allow: owner } ],
        ){
          query: Query
        }
        "#,
        dynaql::AuthConfig {
            allowed_owner_ops: dynaql::Operations::all(),
            oidc_providers: vec![dynaql::OidcProvider {
                issuer: url::Url::parse("https://my.idp.com").unwrap(),
                groups_claim: DEFAULT_GROUPS_CLAIM.to_string(),
            }],
            ..Default::default()
        }
    );

    parse_test!(
        owner_rule_with_ops,
        r#"
        schema @auth(
          providers: [ { type: oidc, issuer: "https://my.idp.com" } ]
          rules: [ { allow: owner, operations: ["create"] } ],
        ){
          query: Query
        }
        "#,
        dynaql::AuthConfig {
            allowed_owner_ops: dynaql::Operations::CREATE,
            oidc_providers: vec![dynaql::OidcProvider {
                issuer: url::Url::parse("https://my.idp.com").unwrap(),
                groups_claim: DEFAULT_GROUPS_CLAIM.to_string(),
            }],
            ..Default::default()
        }
    );

    parse_test!(
        oidc_without_rule,
        r#"
        schema @auth(
          providers: [ { type: oidc, issuer: "https://my.idp.com" } ]
        ){
          query: Query
        }
        "#,
        dynaql::AuthConfig {
            oidc_providers: vec![dynaql::OidcProvider {
                issuer: url::Url::parse("https://my.idp.com").unwrap(),
                groups_claim: DEFAULT_GROUPS_CLAIM.to_string(),
            }],
            ..Default::default()
        }
    );

    parse_fail!(
        oidc_with_missing_field,
        r#"
        schema @auth(
          providers: [ { type: oidc } ]
        ){
          query: Query
        }
        "#,
        "auth provider: missing field `issuer`"
    );

    parse_fail!(
        type_auth_without_model,
        r#"
        type Todo @auth(rules: []) {
          id: ID!
        }
        "#,
        "The @auth directive can only be used on @model types"
    );

    parse_fail!(
        type_auth_with_provider,
        r#"
        type Todo @model @auth(providers: [ { type: oidc, issuer: "https://my.idp.com" } ]) {
          id: ID!
        }
        "#,
        "auth providers can only be configured globally"
    );

    parse_test!(
        issuer_url_from_variable,
        r#"
        schema @auth(
          providers: [ { type: oidc, issuer: "{{ env.ISSUER_URL }}" } ]
        ){
          query: Query
        }
        "#,
        HashMap::from([("ISSUER_URL".to_string(), "https://my.idp.com".to_string())]),
        dynaql::AuthConfig {
            oidc_providers: vec![dynaql::OidcProvider {
                issuer: url::Url::parse("https://my.idp.com").unwrap(),
                groups_claim: DEFAULT_GROUPS_CLAIM.to_string(),
            }],
            ..Default::default()
        }
    );

    parse_fail!(
        issuer_url_from_nonexistent_variable,
        r#"
        schema @auth(
          providers: [ { type: oidc, issuer: "{{ env.ISSUER_URL }}" } ]
        ){
          query: Query
        }
        "#,
        HashMap::new(),
        "undefined variable `ISSUER_URL`"
    );

    parse_fail!(
        issuer_url_from_invalid_template_key,
        r#"
        schema @auth(
          providers: [ { type: oidc, issuer: "{{ ISSUER_URL }}" } ]
        ){
          query: Query
        }
        "#,
        HashMap::new(),
        "auth provider: right now only variables scoped with 'env.' are supported: `ISSUER_URL`"
    );

    parse_fail!(
        issuer_url_empty,
        r#"
        schema @auth(
          providers: [ { type: oidc, issuer: "" } ]
        ){
          query: Query
        }
        "#,
        "OIDC provider: relative URL without a base"
    );

    parse_test!(
        jwt_provider,
        r#"
        schema @auth(
          providers: [ { type: jwt, issuer: "{{ env.ISSUER_URL }}", secret: "{{ env.JWT_SECRET }}" } ]
        ){
          query: Query
        }
        "#,
        HashMap::from([
            ("ISSUER_URL".to_string(), "https://my.idp.com".to_string()),
            ("JWT_SECRET".to_string(), "s3cr3t".to_string())
        ]),
        dynaql::AuthConfig {
            jwt_providers: vec![dynaql::JwtProvider {
                issuer: url::Url::parse("https://my.idp.com").unwrap(),
                groups_claim: DEFAULT_GROUPS_CLAIM.to_string(),
                secret: "s3cr3t".to_string(),
            }],
            ..Default::default()
        }
    );

    parse_fail!(
        multiple_providers,
        r#"
        schema @auth(
          providers: [ { type: foo }, { type: bar } ]
        ){
          query: Query
        }
        "#,
        "only one auth provider can be configured right now"
    );
}
