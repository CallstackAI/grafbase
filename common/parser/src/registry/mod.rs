//! TODO:
//!
//! -> Split each of the creation and add tests with SDL
//!
use case::CaseExt;

use dynaql::indexmap::IndexMap;
use dynaql::registry::enums::DynaqlEnum;
use dynaql::registry::relations::MetaRelation;
use dynaql::registry::transformers::Transformer;
use dynaql::registry::{
    resolvers::dynamo_mutation::DynamoMutationResolver, resolvers::Resolver, resolvers::ResolverType,
    variables::VariableResolveDefinition, MetaField, MetaInputValue, MetaType,
};
use dynaql::registry::{MetaEnumValue, Registry};
use dynaql::validation::dynamic_validators::DynValidator;
use dynaql::AuthConfig;
use dynaql_parser::types::{BaseType, FieldDefinition, ObjectType, TypeDefinition};
use grafbase::auth::Operations;
use std::fmt::Display;

use crate::registry::names::MetaNames;
use crate::rules::length_directive::{LENGTH_DIRECTIVE, MAX_ARGUMENT, MIN_ARGUMENT};
use crate::rules::visitor::VisitorContext;
use crate::utils::{to_input_type, to_lower_camelcase};

mod mutations;
pub mod names;
mod pagination;
mod relations;
mod search;
pub use mutations::{add_mutation_create, add_mutation_update, NumericFieldKind};
pub use pagination::{add_query_paginated_collection, generate_pagination_args};
pub use search::add_query_search;

fn register_dynaql_enum<T: DynaqlEnum>(registry: &mut Registry) -> BaseType {
    let type_name = T::ty().to_string();
    registry.create_type(
        |_| MetaType::Enum {
            name: type_name.clone(),
            description: None,
            enum_values: IndexMap::from_iter(
                T::values()
                    .into_iter()
                    .map(|value| {
                        (
                            value.clone(),
                            MetaEnumValue {
                                name: value,
                                description: None,
                                deprecation: dynaql::registry::Deprecation::NoDeprecated,
                                visible: None,
                                value: None,
                            },
                        )
                    })
                    .collect::<Vec<_>>(),
            ),
            visible: None,
            rust_typename: type_name.clone(),
        },
        &type_name,
        &type_name,
    );
    BaseType::named(&type_name)
}

/// Create an input type for a non_primitive Type.
pub fn add_input_type_non_primitive(ctx: &mut VisitorContext<'_>, object: &ObjectType, type_name: &str) -> String {
    let type_name = type_name.to_string();
    let input_type = format!("{}Input", type_name.to_camel());

    // Input
    ctx.registry.get_mut().create_type(
        |_| MetaType::InputObject {
            name: input_type.clone(),
            description: Some(format!("{type_name} input type.")),
            oneof: false,
            input_fields: {
                let mut input_fields = IndexMap::new();
                for field in &object.fields {
                    let name = &field.node.name.node;

                    input_fields.insert(
                        name.clone().to_string(),
                        MetaInputValue {
                            description: field.node.description.clone().map(|x| x.node),
                            ..MetaInputValue::new(
                                name.to_string(),
                                to_input_type(&ctx.types, field.node.ty.clone().node),
                            )
                        },
                    );
                }
                input_fields
            },
            visible: None,
            rust_typename: input_type.clone(),
        },
        &input_type,
        &input_type,
    );

    input_type
}

/// Add the remove mutation for a given Object
pub fn add_remove_mutation(ctx: &mut VisitorContext<'_>, type_name: &str, auth: Option<&AuthConfig>) {
    let type_name = type_name.to_string();
    let delete_payload_name = format!("{}DeletePayload", type_name.to_camel());

    // DeletePayload
    ctx.registry.get_mut().create_type(
        |_| MetaType::Object {
            name: delete_payload_name.clone(),
            description: None,
            fields: {
                let mut fields = IndexMap::new();
                let name = "deletedId".to_string();
                fields.insert(
                    name.clone(),
                    MetaField {
                        name,
                        description: None,
                        args: Default::default(),
                        // TODO: Should be infered from the entity depending on the directives
                        ty: "ID!".to_string(),
                        deprecation: Default::default(),
                        cache_control: Default::default(),
                        external: false,
                        requires: None,
                        provides: None,
                        visible: None,
                        compute_complexity: None,
                        edges: Vec::new(),
                        relation: None,
                        resolve: None,
                        transformer: Some(Transformer::JSONSelect {
                            property: "id".to_string(),
                        }),
                        plan: None,
                        required_operation: Some(Operations::DELETE),
                        auth: auth.cloned(),
                    },
                );
                fields
            },
            cache_control: Default::default(),
            extends: false,
            keys: None,
            is_node: false,
            visible: None,
            is_subscription: false,
            rust_typename: delete_payload_name.clone(),
            constraints: vec![],
        },
        &delete_payload_name,
        &delete_payload_name,
    );

    // deleteMutation
    ctx.mutations.push(MetaField {
        name: format!("{}Delete", to_lower_camelcase(&type_name)),
        description: Some(format!("Delete a {type_name} by ID or unique field")),
        args: {
            let mut args = IndexMap::new();
            args.insert(
                "by".to_owned(),
                MetaInputValue::new("by", format!("{type_name}ByInput!")),
            );
            args
        },
        ty: delete_payload_name,
        deprecation: dynaql::registry::Deprecation::NoDeprecated,
        cache_control: Default::default(),
        external: false,
        provides: None,
        requires: None,
        visible: None,
        edges: Vec::new(),
        relation: None,
        compute_complexity: None,
        resolve: Some(Resolver {
            id: Some(format!("{}_delete_resolver", type_name.to_lowercase())),
            r#type: ResolverType::DynamoMutationResolver(DynamoMutationResolver::DeleteNode {
                ty: type_name,
                by: VariableResolveDefinition::InputTypeName("by".to_owned()),
            }),
        }),
        plan: None,
        transformer: None,
        required_operation: Some(Operations::DELETE),
        auth: auth.cloned(),
    });
}

fn get_length_validator(field: &FieldDefinition) -> Option<DynValidator> {
    use tuple::Map;
    field
        .directives
        .iter()
        .find(|directive| directive.node.name.node == LENGTH_DIRECTIVE)
        .map(|directive| {
            let (min_value, max_value) = (MIN_ARGUMENT, MAX_ARGUMENT).map(|argument_name| {
                directive.node.get_argument(argument_name).and_then(|argument| {
                    if let dynaql_value::ConstValue::Number(ref min) = argument.node {
                        min.as_u64().and_then(|min| min.try_into().ok())
                    } else {
                        None
                    }
                })
            });
            DynValidator::length(min_value, max_value)
        })
}

/// Used to keep track of the parent relation when created nested input types
/// TODO: Merge it with MetaRelation?
pub struct ParentRelation<'a> {
    /// TypeDefinition of @model type
    model_type_definition: &'a TypeDefinition,
    meta: &'a MetaRelation,
}

impl<'a> Display for ParentRelation<'a> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{} relation of {}",
            self.meta.name,
            MetaNames::model(self.model_type_definition)
        )
    }
}
