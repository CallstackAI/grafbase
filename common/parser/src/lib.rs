#[cfg(test)]
#[macro_use]
extern crate assert_matches;

use std::collections::{HashMap, HashSet};

use dynaql::registry::enums::DynaqlEnums;
use dynaql::registry::scalars::{PossibleScalar, SDLDefinitionScalar};
use dynaql_parser::{parse_schema, Error as ParserError};
use quick_error::quick_error;
use rules::auth_directive::AuthDirective;
use rules::basic_type::BasicType;
use rules::check_field_lowercase::CheckFieldCamelCase;
use rules::check_known_directives::CheckAllDirectivesAreKnown;
use rules::check_type_collision::CheckTypeCollision;
use rules::check_type_validity::CheckTypeValidity;
use rules::check_types_underscore::CheckBeginsWithDoubleUnderscore;
use rules::default_directive::DefaultDirective;
use rules::default_directive_types::DefaultDirectiveTypes;
use rules::directive::Directives;
use rules::enum_type::EnumType;
use rules::extend_query_and_mutation_types::ExtendQueryAndMutationTypes;
use rules::length_directive::LengthDirective;
use rules::model_directive::ModelDirective;
use rules::one_of_directive::OneOfDirective;
use rules::openapi_directive::OpenApiVisitor;
use rules::relations::{relations_rules, RelationEngine};
use rules::resolver_directive::ResolverDirective;
use rules::search_directive::SearchDirective;
use rules::unique_directive::UniqueDirective;
use rules::unique_fields::UniqueObjectFields;
use rules::visitor::{visit, RuleError, Visitor, VisitorContext};

mod models;

use crate::rules::cache_directive::{CacheDirective, CacheVisitor};
pub use dynaql::registry::Registry;
pub use migration_detection::{required_migrations, RequiredMigration};
pub use rules::openapi_directive::{OpenApiDirective, OpenApiQueryNamingStrategy, OpenApiTransforms};

use crate::rules::scalar_hydratation::ScalarHydratation;

mod directive_de;
mod dynamic_string;
mod migration_detection;
mod registry;
mod rules;
#[cfg(test)]
mod tests;
mod utils;

quick_error! {
    #[derive(Debug)]
    pub enum Error {
        Parser(err: ParserError) {
            from()
            source(err)
            display("{}", err)
        }
        Validation(err: Vec<RuleError>) {
            from()
            display("{:?}", err)
        }
    }
}

impl Error {
    #[cfg(test)]
    fn validation_errors(self) -> Option<Vec<RuleError>> {
        if let Error::Validation(err) = self {
            Some(err)
        } else {
            None
        }
    }
}

/// Transform the input schema into a Registry
pub fn to_registry<S: AsRef<str>>(input: S) -> Result<Registry, Error> {
    Ok(to_registry_with_variables(input, &HashMap::new())?.registry)
}

pub struct ParseResult {
    pub registry: Registry,
    pub required_resolvers: HashSet<String>,
    pub openapi_directives: Vec<OpenApiDirective>,
}

/// Transform the input schema into a Registry in the context of provided environment variables
pub fn to_registry_with_variables<S: AsRef<str>>(
    input: S,
    variables: &HashMap<String, String>,
) -> Result<ParseResult, Error> {
    let directives = Directives::new()
        .with::<AuthDirective>()
        .with::<DefaultDirective>()
        .with::<LengthDirective>()
        .with::<ModelDirective>()
        .with::<OneOfDirective>()
        .with::<RelationEngine>()
        .with::<ResolverDirective>()
        .with::<UniqueDirective>()
        .with::<SearchDirective>()
        .with::<OpenApiDirective>()
        .with::<CacheDirective>();

    let mut rules = rules::visitor::VisitorNil
        .with(CheckBeginsWithDoubleUnderscore)
        .with(CheckFieldCamelCase)
        .with(CheckTypeValidity)
        .with(SearchDirective)
        .with(ModelDirective)
        .with(AuthDirective)
        .with(ResolverDirective)
        .with(CacheVisitor)
        .with(BasicType)
        .with(ExtendQueryAndMutationTypes)
        .with(EnumType)
        .with(ScalarHydratation)
        .with(LengthDirective)
        .with(UniqueObjectFields)
        .with(OpenApiVisitor)
        .with(CheckAllDirectivesAreKnown::default());

    let schema = format!(
        "{}\n{}\n{}\n{}",
        input.as_ref(),
        DynaqlEnums::sdl(),
        PossibleScalar::sdl(),
        directives.to_definition(),
    );
    let schema = parse_schema(schema)?;

    let mut ctx = VisitorContext::new_with_variables(&schema, variables);
    // Building all relations first are it requires to parse the whole schema (for ManyToMany). This allows later
    // rules to rely on RelationEngine::get to have correct information on relations.
    visit(
        &mut relations_rules().with(CheckTypeCollision::default()),
        &mut ctx,
        &schema,
    );
    if !ctx.errors.is_empty() {
        return Err(ctx.errors.into());
    }

    visit(&mut rules, &mut ctx, &schema);

    // FIXME: Get rid of the ugly double pass.
    let mut second_pass_rules = rules::visitor::VisitorNil.with(DefaultDirectiveTypes);
    visit(&mut second_pass_rules, &mut ctx, &schema);

    if !ctx.errors.is_empty() {
        return Err(ctx.errors.into());
    }

    let (registry, required_resolvers, openapi_directives) = ctx.finish();
    Ok(ParseResult {
        registry,
        required_resolvers,
        openapi_directives,
    })
}
