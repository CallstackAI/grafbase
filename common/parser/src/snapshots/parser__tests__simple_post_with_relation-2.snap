---
source: common/parser/src/lib.rs
assertion_line: 221
expression: sdl
---
type Author {
	id: ID!
	name: String!
	lastname: String!
	country: Country!
}
"""
Input to create a new Author
"""
input AuthorCreateInput {
	name: String!
	lastname: String!
	country: Country!
}
type AuthorCreatePayload {
	author: Author
}
type AuthorDeletePayload {
	deletedId: ID!
}
type Blog {
	id: ID!
	posts: [Post]
	owner: Author!
}
"""
Input to create a new Blog
"""
input BlogCreateInput {
	posts: [ID]
	owner: ID!
}
type BlogCreatePayload {
	blog: Blog
}
type BlogDeletePayload {
	deletedId: ID!
}
enum Country {
	FRANCE
	NOT_FRANCE
}
type Mutation {
	"""
	Create a Blog
	"""
	blogCreate(input: BlogCreateInput!): BlogCreatePayload
	"""
	Delete a Blog by ID
	"""
	blogDelete(id: ID!): BlogDeletePayload
	"""
	Create a Post
	"""
	postCreate(input: PostCreateInput!): PostCreatePayload
	"""
	Delete a Post by ID
	"""
	postDelete(id: ID!): PostDeletePayload
	"""
	Create a Author
	"""
	authorCreate(input: AuthorCreateInput!): AuthorCreatePayload
	"""
	Delete a Author by ID
	"""
	authorDelete(id: ID!): AuthorDeletePayload
}
type Post {
	id: ID!
	content: String!
	authors: [Author]
}
"""
Input to create a new Post
"""
input PostCreateInput {
	content: String!
	authors: [ID]
}
type PostCreatePayload {
	post: Post
}
type PostDeletePayload {
	deletedId: ID!
}
type Query {
	"""
	Get a Blog by ID
	"""
	blog(id: ID!): Blog
	"""
	Unpaginated query to fetch the whole list of `Blog`.
	"""
	blogCollection: [Blog]
	"""
	Get a Post by ID
	"""
	post(id: ID!): Post
	"""
	Unpaginated query to fetch the whole list of `Post`.
	"""
	postCollection: [Post]
	"""
	Get a Author by ID
	"""
	author(id: ID!): Author
	"""
	Unpaginated query to fetch the whole list of `Author`.
	"""
	authorCollection: [Author]
}
schema {
	query: Query
	mutation: Mutation
}

